File: project/backend-services/rule-management-backend/domain/src/main/java/pt/sensae/services/rule/management/backend/domain/RuleScenario.java
package pt.sensae.services.rule.management.backend.domain;

import java.util.Objects;

public record RuleScenario(RuleScenarioId id, RuleScenarioContent content, RuleScenarioOwners owners,
                           RuleApplied isApplied, RuleDeleted isDeleted) {

    public RuleScenario withContent(RuleScenarioContent content) {
        return new RuleScenario(id, content, owners, new RuleApplied(false), isDeleted);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        RuleScenario that = (RuleScenario) o;

        return Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return id != null ? id.hashCode() : 0;
    }

}

File: project/frontend-services/libs/rule-management/model/src/lib/model/RuleScenario.ts
import {RuleScenarioContent} from "./RuleScenarioContent";
import {RuleScenarioId} from "./RuleScenarioId";
import {IsDraft} from "./IsDraft";

export class RuleScenario {
  constructor(
    public id: RuleScenarioId,
    public content: RuleScenarioContent,
    public isDraft: IsDraft
  ) {
  }

  static empty() {
    return new RuleScenario(
      RuleScenarioId.empty(),
      RuleScenarioContent.empty(),
      IsDraft.default()
    );
  }

  copy() {
    return new RuleScenario(this.id.copy(), this.content.copy(), new IsDraft(true));
  }

  isValid() {
    return (
      this.id.value.trim().length !== 0 &&
      this.content.value.trim().length !== 0
    );
  }
}
